crates/shared/src/sources/balancer_v2/graph_api.rs

```diff
@@ -114,56 +115,80 @@ pub struct RegisteredPools {
     /// weighted pools can be considered up to date.
     pub fetched_block_number: u64,
     /// The registered Pools
-    pub weighted_pools_by_factory: HashMap<H160, Vec<RegisteredWeightedPool>>,
-    pub stable_pools_by_factory: HashMap<H160, Vec<RegisteredStablePool>>,
+    pub pools: Vec<PoolData>,
 }
 
-impl RegisteredPools {
-    fn from_pool_data(fetched_block_number: u64, pools: Vec<PoolData>) -> Result<Self> {
-        let mut registered_pools = Self {
-            fetched_block_number,
-            weighted_pools_by_factory: Default::default(),
-            stable_pools_by_factory: Default::default(),
-        };
+/// Pool data from the Balancer V2 subgraph.
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct PoolData {
+    #[serde(rename = "poolType")]
+    pub pool_type: PoolType,
+    pub id: H256,
+    pub address: H160,
+    pub factory: H160,
+    pub tokens: Vec<Token>,
+}
 
-        for pool in pools {
-            let common = CommonPoolData {
-                pool_id: pool.id,
-                pool_address: pool.address,
-                tokens: pool.tokens.iter().map(|token| token.address).collect(),
-                scaling_exponents: pool
-                    .tokens
-                    .iter()
-                    .map(|token| scaling_exponent_from_decimals(token.decimals))
-                    .collect::<Result<_>>()?,
-                block_created: fetched_block_number,
-            };
-            match pool.pool_type {
-                PoolType::Weighted => registered_pools
-                    .weighted_pools_by_factory
-                    .entry(pool.factory)
-                    .or_default()
-                    .push(RegisteredWeightedPool {
-                        common,
-                        normalized_weights: pool
-                            .tokens
-                            .iter()
-                            .map(|token| {
-                                token.weight.ok_or_else(|| {
-                                    anyhow!("missing weights for pool {:?}", pool.id)
-                                })
-                            })
-                            .collect::<Result<_>>()?,
-                    }),
-                PoolType::Stable => registered_pools
-                    .stable_pools_by_factory
-                    .entry(pool.factory)
-                    .or_default()
-                    .push(RegisteredStablePool { common }),
-            }
-        }
+/// Supported pool kinds.
+#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Hash)]
+pub enum PoolType {
+    Stable,
+    Weighted,
+}
+
+/// Token data for pools.
+#[serde_as]
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct Token {
+    pub address: H160,
+    pub decimals: u8,
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub weight: Option<Bfp>,
```

Comment:
Would it be safe to assume even weights for pools in which this is None (may allow more consolidation of the data model)?

Reply:
The risk is we may assume we have weights if they are missing. They used to be on Rinkeby for example (but because of subgraph updates they are available now).

We convert the `None` into an `Err` for a weighted pool in those cases.